packages/flutter_tools/lib/src/asset.dart

```diff
@@ -613,7 +622,9 @@ class _AssetDirectoryCache {
       final List<String> paths = <String>[];
       for (final FileSystemEntity entity in globals.fs.directory(directory).listSync(recursive: true)) {
         final String path = entity.path;
-        if (globals.fs.isFileSync(path) && !_excluded.any((String exclude) => path.startsWith(exclude))) {
+        if (globals.fs.isFileSync(path)
+          && assetPath != path
```

Comment:
need to check the assetPath != path, otherwise the tool recognizes `assets: - foo.txt` as its own asset variant.